File: src/main/java/no/novari/flyt/egrunnerverv/gateway/configuration/ResourceEntityConsumersConfiguration.java
package no.novari.flyt.egrunnerverv.gateway.configuration;

import jakarta.annotation.Nullable;
import no.fint.model.resource.FintLinks;
import no.fint.model.resource.administrasjon.personal.PersonalressursResource;
import no.fint.model.resource.arkiv.kodeverk.JournalStatusResource;
import no.fint.model.resource.arkiv.kodeverk.JournalpostTypeResource;
import no.fint.model.resource.arkiv.kodeverk.SkjermingshjemmelResource;
import no.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource;
import no.fint.model.resource.arkiv.noark.AdministrativEnhetResource;
import no.fint.model.resource.arkiv.noark.ArkivressursResource;
import no.fint.model.resource.felles.PersonResource;
import no.novari.cache.FintCache;
import no.novari.flyt.egrunnerverv.gateway.ResourceLinkUtil;
import no.novari.flyt.egrunnerverv.gateway.instance.ResourceRepository;
import no.novari.kafka.consuming.ErrorHandlerConfiguration;
import no.novari.kafka.consuming.ErrorHandlerFactory;
import no.novari.kafka.consuming.ListenerConfiguration;
import no.novari.kafka.consuming.ParameterizedListenerContainerFactoryService;
import no.novari.kafka.topic.name.EntityTopicNameParameters;
import no.novari.kafka.topic.name.TopicNamePrefixParameters;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.listener.ConcurrentMessageListenerContainer;

import java.util.function.Consumer;

@Configuration
public class ResourceEntityConsumersConfiguration {

    private final ParameterizedListenerContainerFactoryService listenerFactoryService;
    private final ErrorHandlerFactory errorHandlerFactory;

    public ResourceEntityConsumersConfiguration(
            ParameterizedListenerContainerFactoryService listenerFactoryService,
            ErrorHandlerFactory errorHandlerFactory
    ) {
        this.listenerFactoryService = listenerFactoryService;
        this.errorHandlerFactory = errorHandlerFactory;
    }

    private <T extends FintLinks> ConcurrentMessageListenerContainer<String, T> createCacheConsumer(
            String resourceName,
            Class<T> resourceClass,
            FintCache<String, T> cache,
            @Nullable Consumer<T> afterCache
    ) {
        ListenerConfiguration listenerConfig = ListenerConfiguration.stepBuilder()
                .groupIdApplicationDefault()
                .maxPollRecordsKafkaDefault()
                .maxPollIntervalKafkaDefault()
                .seekToBeginningOnAssignment()
                .build();

        return listenerFactoryService
                .createRecordListenerContainerFactory(
                        resourceClass,
                        record -> {
                            T value = record.value();
                            cache.put(ResourceLinkUtil.getSelfLinks(value), value);
                            if (afterCache != null) {
                                afterCache.accept(value);
                            }
                        },
                        listenerConfig,
                        errorHandlerFactory.createErrorHandler(ErrorHandlerConfiguration.<T>builder().build())
                )
                .createContainer(
                        EntityTopicNameParameters.builder()
                                .topicNamePrefixParameters(
                                        TopicNamePrefixParameters.stepBuilder()
                                                .orgIdApplicationDefault()
                                                .domainContextApplicationDefault()
                                                .build()
                                )
                                .resourceName(resourceName)
                                .build()
                );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, AdministrativEnhetResource> administrativEnhetResourceEntityConsumer(
            FintCache<String, AdministrativEnhetResource> administrativEnhetResourceCache
    ) {
        return createCacheConsumer(
                "arkiv-noark-administrativenhet",
                AdministrativEnhetResource.class,
                administrativEnhetResourceCache,
                null
        );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, ArkivressursResource> arkivressursResourceEntityConsumer(
            FintCache<String, ArkivressursResource> arkivressursResourceCache,
            ResourceRepository resourceRepository
    ) {
        return createCacheConsumer(
                "arkiv-noark-arkivressurs",
                ArkivressursResource.class,
                arkivressursResourceCache,
                resourceRepository::updateArkivRessurs
        );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceEntityConsumer(
            FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache
    ) {
        return createCacheConsumer(
                "arkiv-kodeverk-tilgangsrestriksjon",
                TilgangsrestriksjonResource.class,
                tilgangsrestriksjonResourceCache,
                null
        );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, SkjermingshjemmelResource> skjermingshjemmelResourceEntityConsumer(
            FintCache<String, SkjermingshjemmelResource> skjermingshjemmelResourceCache
    ) {
        return createCacheConsumer(
                "arkiv-kodeverk-skjermingshjemmel",
                SkjermingshjemmelResource.class,
                skjermingshjemmelResourceCache,
                null
        );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, JournalStatusResource> journalstatusResourceEntityConsumer(
            FintCache<String, JournalStatusResource> journalStatusResourceCache
    ) {
        return createCacheConsumer(
                "arkiv-kodeverk-journalstatus",
                JournalStatusResource.class,
                journalStatusResourceCache,
                null
        );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, JournalpostTypeResource> journalposttypeEntityConsumer(
            FintCache<String, JournalpostTypeResource> journalpostTypeResourceCache
    ) {
        return createCacheConsumer(
                "arkiv-kodeverk-journalposttype",
                JournalpostTypeResource.class,
                journalpostTypeResourceCache,
                null
        );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, PersonalressursResource> personalressursResourceEntityConsumer(
            FintCache<String, PersonalressursResource> personalressursResourceCache,
            ResourceRepository resourceRepository
    ) {
        return createCacheConsumer(
                "administrasjon-personal-personalressurs",
                PersonalressursResource.class,
                personalressursResourceCache,
                resourceRepository::updatePersonalRessurs
        );
    }

    @Bean
    ConcurrentMessageListenerContainer<String, PersonResource> personResourceEntityConsumer(
            FintCache<String, PersonResource> personResourceCache
    ) {
        return createCacheConsumer(
                "administrasjon-personal-person",
                PersonResource.class,
                personResourceCache,
                null
        );
    }
}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/configuration/ResourceEntityConsumersConfiguration.kt
package no.novari.flyt.egrunnerverv.gateway.configuration

import jakarta.annotation.Nullable
import no.fint.model.resource.FintLinks
import no.fint.model.resource.administrasjon.personal.PersonalressursResource
import no.fint.model.resource.arkiv.kodeverk.JournalStatusResource
import no.fint.model.resource.arkiv.kodeverk.JournalpostTypeResource
import no.fint.model.resource.arkiv.kodeverk.SkjermingshjemmelResource
import no.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource
import no.fint.model.resource.arkiv.noark.AdministrativEnhetResource
import no.fint.model.resource.arkiv.noark.ArkivressursResource
import no.fint.model.resource.felles.PersonResource
import no.novari.cache.FintCache
import no.novari.flyt.egrunnerverv.gateway.ResourceLinkUtil
import no.novari.flyt.egrunnerverv.gateway.instance.ResourceRepository
import no.novari.kafka.consuming.ErrorHandlerConfiguration
import no.novari.kafka.consuming.ErrorHandlerFactory
import no.novari.kafka.consuming.ListenerConfiguration
import no.novari.kafka.consuming.ParameterizedListenerContainerFactoryService
import no.novari.kafka.topic.name.EntityTopicNameParameters
import no.novari.kafka.topic.name.TopicNamePrefixParameters
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.listener.ConcurrentMessageListenerContainer
import java.util.function.Consumer

@Configuration
class ResourceEntityConsumersConfiguration(
    private val listenerFactoryService: ParameterizedListenerContainerFactoryService,
    private val errorHandlerFactory: ErrorHandlerFactory,
) {
    private fun <T : FintLinks> createCacheConsumer(
        resourceName: String,
        resourceClass: Class<T>,
        cache: FintCache<String, T>,
        @Nullable afterCache: Consumer<T>?,
    ): ConcurrentMessageListenerContainer<String, T> {
        val listenerConfig =
            ListenerConfiguration
                .stepBuilder()
                .groupIdApplicationDefault()
                .maxPollRecordsKafkaDefault()
                .maxPollIntervalKafkaDefault()
                .seekToBeginningOnAssignment()
                .build()

        val errorHandler =
            errorHandlerFactory.createErrorHandler(
                ErrorHandlerConfiguration
                    .stepBuilder<T>()
                    .noRetries()
                    .skipFailedRecords()
                    .build(),
            )

        return listenerFactoryService
            .createRecordListenerContainerFactory<T>(
                resourceClass,
                { record: ConsumerRecord<String, T> ->
                    val value = record.value()
                    cache.put(ResourceLinkUtil.getSelfLinks(value), value)
                    afterCache?.accept(value)
                },
                listenerConfig,
                errorHandler,
            ).createContainer(
                EntityTopicNameParameters
                    .builder()
                    .topicNamePrefixParameters(
                        TopicNamePrefixParameters
                            .stepBuilder()
                            .orgIdApplicationDefault()
                            .domainContextApplicationDefault()
                            .build(),
                    ).resourceName(resourceName)
                    .build(),
            )
    }

    @Bean
    fun administrativEnhetResourceEntityConsumer(
        administrativEnhetResourceCache: FintCache<String, AdministrativEnhetResource>,
    ): ConcurrentMessageListenerContainer<String, AdministrativEnhetResource> =
        createCacheConsumer(
            resourceName = "arkiv-noark-administrativenhet",
            resourceClass = AdministrativEnhetResource::class.java,
            cache = administrativEnhetResourceCache,
            afterCache = null,
        )

    @Bean
    fun arkivressursResourceEntityConsumer(
        arkivressursResourceCache: FintCache<String, ArkivressursResource>,
        resourceRepository: ResourceRepository,
    ): ConcurrentMessageListenerContainer<String, ArkivressursResource> =
        createCacheConsumer(
            resourceName = "arkiv-noark-arkivressurs",
            resourceClass = ArkivressursResource::class.java,
            cache = arkivressursResourceCache,
            afterCache = resourceRepository::updateArkivRessurs,
        )

    @Bean
    fun tilgangsrestriksjonResourceEntityConsumer(
        tilgangsrestriksjonResourceCache: FintCache<String, TilgangsrestriksjonResource>,
    ): ConcurrentMessageListenerContainer<String, TilgangsrestriksjonResource> =
        createCacheConsumer(
            resourceName = "arkiv-kodeverk-tilgangsrestriksjon",
            resourceClass = TilgangsrestriksjonResource::class.java,
            cache = tilgangsrestriksjonResourceCache,
            afterCache = null,
        )

    @Bean
    fun skjermingshjemmelResourceEntityConsumer(
        skjermingshjemmelResourceCache: FintCache<String, SkjermingshjemmelResource>,
    ): ConcurrentMessageListenerContainer<String, SkjermingshjemmelResource> =
        createCacheConsumer(
            resourceName = "arkiv-kodeverk-skjermingshjemmel",
            resourceClass = SkjermingshjemmelResource::class.java,
            cache = skjermingshjemmelResourceCache,
            afterCache = null,
        )

    @Bean
    fun journalstatusResourceEntityConsumer(
        journalStatusResourceCache: FintCache<String, JournalStatusResource>,
    ): ConcurrentMessageListenerContainer<String, JournalStatusResource> =
        createCacheConsumer(
            resourceName = "arkiv-kodeverk-journalstatus",
            resourceClass = JournalStatusResource::class.java,
            cache = journalStatusResourceCache,
            afterCache = null,
        )

    @Bean
    fun journalposttypeEntityConsumer(
        journalpostTypeResourceCache: FintCache<String, JournalpostTypeResource>,
    ): ConcurrentMessageListenerContainer<String, JournalpostTypeResource> =
        createCacheConsumer(
            resourceName = "arkiv-kodeverk-journalposttype",
            resourceClass = JournalpostTypeResource::class.java,
            cache = journalpostTypeResourceCache,
            afterCache = null,
        )

    @Bean
    fun personalressursResourceEntityConsumer(
        personalressursResourceCache: FintCache<String, PersonalressursResource>,
        resourceRepository: ResourceRepository,
    ): ConcurrentMessageListenerContainer<String, PersonalressursResource> =
        createCacheConsumer(
            resourceName = "administrasjon-personal-personalressurs",
            resourceClass = PersonalressursResource::class.java,
            cache = personalressursResourceCache,
            afterCache = resourceRepository::updatePersonalRessurs,
        )

    @Bean
    fun personResourceEntityConsumer(
        personResourceCache: FintCache<String, PersonResource>,
    ): ConcurrentMessageListenerContainer<String, PersonResource> =
        createCacheConsumer(
            resourceName = "administrasjon-personal-person",
            resourceClass = PersonResource::class.java,
            cache = personResourceCache,
            afterCache = null,
        )
}
